packages/react/xr/src/hooks.ts

```diff
@@ -102,3 +104,132 @@ export function useSessionModeSupported(mode: XRSessionMode, onError?: (error: a
 export function useSessionFeatureEnabled(feature: string) {
   return useXR(({ session }) => session?.enabledFeatures?.includes(feature) ?? false)
 }
+
+export interface ControllerLocomotionTranslationOptions {
+  speed?: number
+}
+export interface ControllerLocomotionRotationOptions {
+  numberOfDegreesToSnapBy?: number
+  viewControlDeadZone?: number
+  disableSnapTurning?: boolean
+  enableSmoothTurning?: boolean
+  smoothTurningSpeed?: number
+}
+
+/**
+ * A hook for handling basic locomotion in VR
+ * @param options Options that can be provided to customize the locomotion behavior
+ * @returns A ref to be assigned to the <XROrigin> component (i.e. <XROrigin ref={locomotionRef}>)
+ */
+export function useControllerLocomotion(
+  target:
+    | RefObject<Group>
+    | ((velocity: Vector3, rotationVelocity: Euler, deltaTime: number, state: RootState, frame?: XRFrame) => void),
+  translationOptions?: ControllerLocomotionTranslationOptions,
+  rotationOptions?: ControllerLocomotionRotationOptions,
+  movementController?: XRHandedness,
+) {
+  const defaultSpeed = 2
```

Comment:
these can be declared outside the function

Reply:
Wasn't sure how you wanted things to be scoped, so I left them in. Moved them out 👍